question5/main.c

```diff
@@ -0,0 +1,200 @@
+  int segment_id;
+  char roadway_name[NAME_FIELD_WIDTH];
+  char from[NAME_FIELD_WIDTH];
+  char to[NAME_FIELD_WIDTH];
+  enum direction direction;
+  date date;
+  int hour;
+  int traffic;
+} info;
+
+typedef struct infoset {
+  info *info_array;
+  int size;
+  int capacity;
+} infoset;
+
+infoset *new_infoset();
+void free_infoset(infoset *);
+void infoset_push(infoset *, info);
+infoset *infoset_from_csv(FILE *csv);
+void infoset_dump(infoset *);
+int busiest_hour(infoset *);
+int average_traffic(infoset *s, const char *street_name, int hour);
+
+int main(int argc, char **argv) {
+  if (argc != 2) {
+    printf("Usage: trafficanalysis <filename>\n");
+    return 1;
+  } else {
+    FILE *file = fopen(argv[1], "r");
+    if (file != NULL) {
+      infoset *set = infoset_from_csv(file);
+      printf("%d:00 is the busiest hour.\n", busiest_hour(set));
+      printf("Average traffic of HUGUENOT AVE @ 15:00 is %d.\n",
+             average_traffic(set, "HUGUENOT AVE", 15));
+      free_infoset(set);
+    }
+  }
+}
+
+infoset *new_infoset() {
+  infoset *s = malloc(sizeof(infoset));
+  s->info_array = malloc(sizeof(info));
+  s->size = 0;
+  s->capacity = 1;
+
+  return s;
+}
+
+void free_infoset(infoset *s) {
+  free(s->info_array);
+  free(s);
+}
+
+void infoset_push(infoset *s, info inf) {
+  if (s->size < s->capacity) {
+    // Info array has space available.
+    s->info_array[s->size++] = inf;
+  } else {
+    // Double the capacity and retry.
+    info *temp_info_array = malloc(sizeof(info) * s->capacity * 2);
+    int i;
+    for (i = 0; i < s->size; i++) {
+      temp_info_array[i] = s->info_array[i];
+    }
+    free(s->info_array);
+    s->info_array = temp_info_array;
+    s->capacity *= 2;
+
+    infoset_push(s, inf);
+  }
+}
+
+infoset *infoset_from_csv(FILE *csv) {
+  infoset *ret = new_infoset();
+
+  char line[MAX_LINE_LENGTH];
+  // Ignore Header.
+  fgets(line, MAX_LINE_LENGTH, csv);
+  while (fgets(line, MAX_LINE_LENGTH, csv) != NULL) {
+    char *cols[NUM_COLS];
+    int i, j;
+    for (i = 0, j = 1, cols[0] = line; i < MAX_LINE_LENGTH && j < NUM_COLS;
+         i++) {
+      if (line[i] == ',') {
+        line[i] = '\0';
+        cols[j++] = &(line[i + 1]);
+      }
+    }
+
+    // Push into infoset.
+    int segment_id = atoi(cols[1]);
+    char roadway_name[NAME_FIELD_WIDTH];
+    strncpy(roadway_name, cols[2], NAME_FIELD_WIDTH);
+    char from[NAME_FIELD_WIDTH];
+    strncpy(from, cols[3], NAME_FIELD_WIDTH);
+    char to[NAME_FIELD_WIDTH];
+    strncpy(to, cols[4], NAME_FIELD_WIDTH);
+    enum direction direction;
+    if (strcmp("NB", cols[5]) == 0) {
+      direction = NB;
+    } else if (strcmp("SB", cols[5]) == 0) {
+      direction = SB;
+    } else if (strcmp("EB", cols[5]) == 0) {
+      direction = EB;
+    } else {
+      direction = WB;
+    }
+    date date;
+    cols[6][2] = '\0';
+    cols[6][5] = '\0';
+    date.month = atoi(cols[6]);
+    date.day = atoi(cols[6] + 3);
+    date.year = atoi(cols[6] + 6);
```

Comment:
be careful of magic numbers; they can be dangerous if you don't have proper const/#defines and comments